### PR TITLE
Add compiled optimizer + lr scheduler tutorial

### PR DESCRIPTION
**Pull request by mlazos**
Adding a tutorial demonstrating how to compose the LRScheduler with a compiled optimizer

### CODE REVIEW

File: recipes_source/compiling_optimizer_lr_scheduler.rst [hunk 1]
```diff
@@ -0,0 +1,125 @@
+(beta) Running the compiled optimizer with an LR Scheduler
+==========================================================================================
+
+**Author:** `Michael Lazos <https://github.com/mlazos>`_
+
+The optimizer is a key algorithm for training any deep learning model.
+In this example, we will show how to pair the ``torch.compile``d optimizer
```
Comment (svekars) [inline recipes_source/compiling_optimizer_lr_scheduler.rst:7]: Can we not say ``torch.compile``d since it's not really a verb. Maybe something like:
```suggestion
In this example, we will show how to pair the an optimizer, which has been compiled using ``torch.compile``,
```

File: recipes_source/compiling_optimizer_lr_scheduler.rst [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+with the LR schedulers to accelerate training convergence
```
Comment (svekars) [inline recipes_source/compiling_optimizer_lr_scheduler.rst:8]:
```suggestion
with the LR schedulers to accelerate training convergence.
```

File: recipes_source/compiling_optimizer_lr_scheduler.rst [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+
+   import torch
+
+   model = torch.nn.Sequential(
+       *[torch.nn.Linear(1024, 1024, False, device="cuda") for _ in range(10)]
+   )
+   input = torch.rand(1024, device="cuda")
+   output = model(input)
+   output.sum().backward()
+
```
Comment (svekars) [inline recipes_source/compiling_optimizer_lr_scheduler.rst:28]: Can you add a brief description: The code above ... (does xyz)

File: recipes_source/compiling_optimizer_lr_scheduler.rst [hunk 4]
```diff
@@ -0,0 +1,125 @@
...
+Setting up and running the compiled optimizer with LR Scheduler
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
Comment (svekars) [inline recipes_source/compiling_optimizer_lr_scheduler.rst:30]:
```suggestion
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

File: recipes_source/compiling_optimizer_lr_scheduler.rst [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+In this example, we'll use the Adam optimizer with ConstantLR Scheduler
+and create a helper function to wrap the step()
```
Comment (svekars) [inline recipes_source/compiling_optimizer_lr_scheduler.rst:32]:
```suggestion
and create a helper function to wrap the ``step()``
```

**Review by svekars — COMMENTED** [review]
Some editorial suggestions. Thank you!

File: recipes_source/compiling_optimizer_lr_scheduler.rst [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+>> tensor(0.0100)
+
+Extension: What happens with a non-tensor LR?
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+For the curious, we will show how to peek into what happens with ``torch.compile`` when we don't wrap the
+LR in a tensor.
+
+.. code-block:: python
+   # exit cleanly if we are on a device that doesn't support torch.compile
+   if torch.cuda.get_device_capability() < (7, 0):
```
Comment (williamwen42) [inline recipes_source/compiling_optimizer_lr_scheduler.rst:80]: Don't need this block again?

**Review by williamwen42 — COMMENTED** [review]
Will you have a `.py` version up for next release?

**Comment by mlazos** [comment]
> Will you have a `.py` version up for next release?

I want to, but last time we did this we had a strange issue in the docs CI due to a previous test modifying the decomps and causing it to fail, so I went with RST both with this and the previous compiled optimizer tutorial.

**Comment by mlazos** [comment]
> > Will you have a `.py` version up for next release?
> 
> I want to, but last time we did this we had a strange issue in the docs CI due to a previous test modifying the decomps and causing it to fail, so I went with RST both with this and the previous compiled optimizer tutorial.

Talked to svetlana offline, this should be possible

File: recipes_source/compiling_optimizer_lr_scheduler.rst (hunk 6 above)
Comment (williamwen42) [reply recipes_source/compiling_optimizer_lr_scheduler.rst:80]: This wasn't removed in the rst

Comment (mlazos) [reply recipes_source/compiling_optimizer_lr_scheduler.rst:80]: Oh meant to remove the rst entirely, will do that quick

Comment (mlazos) [reply recipes_source/compiling_optimizer_lr_scheduler.rst]: Done

**Review by williamwen42 — APPROVED** [review]

File: recipes_source/compiling_optimizer_lr_scheduler.py [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+"""
+
+#########################################################
+# The optimizer is a key algorithm for training any deep learning model.
+# In this example, we will show how to pair the an optimizer, which has been compiled using ``torch.compile``,
+# with the LR schedulers to accelerate training convergence.
+#
+# .. note::
+#
+#    This tutorial requires PyTorch 2.2.0 or later.
```
Comment (janeyx99) [inline recipes_source/compiling_optimizer_lr_scheduler.py:14]:
```suggestion
#    This tutorial requires PyTorch 2.4.0 or later.
```
The LRSched changes are in 2.4, right?

File: recipes_source/compiling_optimizer_lr_scheduler.py [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+#    ``torch.compile`` is only supported on CUDA devices that have a compute capability of 7.0 or higher.
+
+
+# exit cleanly if we are on a device that doesn't support torch.compile
+if torch.cuda.get_device_capability() < (7, 0):
+    print("Exiting because torch.compile is not supported on this device.")
+    import sys
+    sys.exit(0)
+
+# !!! IMPORTANT !!! Wrap the lr in a tensor if we are pairing the
```
Comment (janeyx99) [inline recipes_source/compiling_optimizer_lr_scheduler.py:55]:
```suggestion
# !!! IMPORTANT !!! Wrap the lr in a Tensor if we are pairing the
```

nit, but maybe refer to Tensor as the class

File: recipes_source/compiling_optimizer_lr_scheduler.py [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+opt = torch.optim.Adam(model.parameters(), lr=0.01)
+sched = torch.optim.lr_scheduler.LinearLR(opt, total_iters=5)
+
+@torch.compile(fullgraph=False)
+def fn():
+    opt.step()
+    sched.step()
+
+# Setup logging to view recompiles
+torch._logging.set_logs(recompiles=True)
```
Comment (janeyx99) [inline recipes_source/compiling_optimizer_lr_scheduler.py:114]: nice advertising for logging haha

**Review by janeyx99 — COMMENTED** [review]
Looks like the build isn't passing, but contentwise this LGTM

File: recipes_source/compiling_optimizer_lr_scheduler.py [hunk 10]
```diff
@@ -0,0 +1,144 @@
...
+########################################################################
+# Sample Output:
+#
+# >> tensor(0.0047)
+# >> tensor(0.0060)
+# >> tensor(0.0073)
+# >> tensor(0.0087)
+# >> tensor(0.0100)
+
+######################################################################
```
Comment (svekars) [inline recipes_source/compiling_optimizer_lr_scheduler.py:82]: conclusion should be the last section

**Review by svekars — APPROVED** [review]
